packages/styled-components/src/test/types.tsx

```diff
@@ -266,12 +284,53 @@ const StyledDiv = styled.div``;
 
 const CustomComponent = (({ ...props }) => {
   return <StyledDiv {...props} />;
-}) as IStyledComponent<'web', 'div'>;
+}) as IStyledComponent<'web', 'div', {}>;
 
 const StyledCustomComponent = styled(CustomComponent)``;
 
 <StyledCustomComponent className="class" />;
 
+const CustomComponentWithRequiredProps: React.FC<{
+  foo: 'red' | 'black';
+  bar?: boolean;
+  hello: (a: string) => string;
+}> = ({ foo, hello, bar }) => (
+  <StyledDiv
+    style={{ backgroundColor: foo, color: bar ? 'blue' : 'green' }}
+    onClick={() => hello('world')}
+  />
+);
+
+const StyledCustomComponentWithRequiredProps = styled(CustomComponentWithRequiredProps)``;
+
+// foo should be typed 'red' | 'black'
+// hello should be typed (a: string) => string
+// bar should be optional
+<StyledCustomComponentWithRequiredProps foo='red' hello={(world) => world} />;
+
+// @ts-expect-error all props have type mismatches
+<StyledCustomComponentWithRequiredProps foo='blue' bar="baz" hello={() => false} />;
+
+// foo should be typed 'red' | 'black'
+// hello should be typed (a: string) => string
+// bar should be typed boolean
+const StyledCustomComponentWithAttrs = styled(
+  CustomComponentWithRequiredProps
+).attrs((props) => ({
+  foo: props.foo === 'black' ? 'red' : 'black',
+  hello: (world) => props.hello(world),
+  bar: !props.bar,
+}));
+
+const StyledCustomComponentWithBadAttrs = styled(
```

Comment:
this needs a @ts-expect-error comment

Reply:
Fixed this, and also updated the negative test case examples